### PR TITLE
Thread-safe samza BeamProducer

### PR DESCRIPTION
**Pull request by logarithm**
Since we use samza 0.12 which might work in multi-threading environment druid producer should be thread safe. `Tranquilizer`'s methods `send` and `flush` already thread safe, but `BeamProducer` was not.

This patch addresses it and makes `BeamProducer` thread safe.

### CODE REVIEW

File: samza/src/main/scala/com/metamx/tranquility/samza/BeamProducer.scala [hunk 1]
```diff
@@ -55,16 +56,18 @@ class BeamProducer(
     throwOnError
   )
 
+  // PermanentMap supports only get by key, but we need to be able to iterate through all streams
+  private val streams = mutable.Set[String]()
```
Comment (gianm) [inline samza/src/main/scala/com/metamx/tranquility/samza/BeamProducer.scala]: This isn't thread safe, does it need to be?

File: samza/src/main/scala/com/metamx/tranquility/samza/BeamProducer.scala [hunk 2]
```diff
@@ -76,6 +79,7 @@ class BeamProducer(
     val sender = senders.getOrElseUpdate(
       streamName, {
         log.info("Creating beam for stream[%s.%s].", systemName, streamName)
+        streams += streamName
```
Comment (gianm) [inline samza/src/main/scala/com/metamx/tranquility/samza/BeamProducer.scala]: This won't be atomic with the update to `senders`, does it need to be?

**Comment by logarithm** [comment]
@gianm Updated with double-checked locking and ConcurrentHashMap

**Comment by gianm** [comment]
LGTM, thanks @logarithm

**Comment by gianm** [comment]
LGTM, thanks @logarithm

**Pull request merged**